storage/innobase/fil/fil0crypt.cc

```diff
@@ -711,8 +711,8 @@ static dberr_t fil_space_decrypt_for_non_full_checksum(
 	/* Calculate the offset where decryption starts */
 	const byte* src = src_frame + header_len;
 	byte* dst = tmp_frame + header_len;
```

Comment:
and here in the `encryption.innodb_import,encryption` test (when no page compression is used) encryption is called with `src == tmp` !
While we can adjust the assert to handle this, I think InnoDB shouldn't abuse the API like that.
@dr-m, what do you think?

Reply:
I found two calls that use the same source and destination buffer and changed them to allocate a temporary buffer instead https://github.com/MariaDB/server/pull/2438/commits/7237813317fe46256f0bcc2191a2e273acee6089. It doesn't look like the calls happen in any hot loops so the allocation overhead probably isn't too bad.

Also added an assertion to make sure they don't overlap and make the failure more explicit https://github.com/MariaDB/server/pull/2438/commits/7ea78661bc409556e986c912217805206f83950c. I'll adjust these changes as needed.

All tests pass on my local. It seems like there sometimes are a few spurious failures on CI, I'm not sure where they're coming from.

Reply:
@tgross35 Thank you. It does not look too bad, only affecting `ALTER TABLE…IMPORT TABLESPACE`. But could you please move the allocation of the temporary buffer up one level, to `fil_tablespace_iterate()`? That is where we allocate `iter.crypt_io_buffer` as well. Could we extend `fil_iterator_t` with a new field `crypt_tmp_buffer`?

Reply:
@dr-m I added a field to `fil_iterator_t` to help `FetchIndexRootPages::run` and updated the signature for `decrypt_decompress` to help `handle_instant_metadata`. I notice there don't seem to be any checks for alloc null/OOM - is this handled automatically somehow?

Reply:
For `aligned_malloc()` some explicit OOM handling could indeed be added fairly easily. InnoDB used to be very unforgiving with regard to memory allocations, killing the server whenever running out of memory.